### ISSUE TITLE
Atomic performance issue when using the buffer-accessor model

### ISSUE THREAD

**Issue by sbalint98** [AdaptiveCpp/AdaptiveCpp#729]
**Bug summary**
When targetting the AMD backend hipSYCL generates inefficient atomic instructions when the buffer-accessor interface is used. In case of the USM interface, the performance is as expected.

**To Reproduce**

The issue is identified on gfx908. In case memory is accessed through an accessor like in this example:
```
    q.submit([&](sycl::handler& h) {
        sycl::accessor X(d_X,h,sycl::read_write);
        h.parallel_for<class axpy>(sycl::range<1>{NUM_THREADS}, [=] (sycl::id<1> it) {
            sycl::atomic_ref<float, sycl::memory_order::relaxed, sycl::memory_scope::device, sycl::access::address_space::global_space> ref(X[0]);
            float one = 1;
            for(int i = 0; i < NUM_ITER; ++i){
                ref.fetch_add(one);
            }
        });
    });
```
The generated code is going to be very slow. For `NUM_THREADS=1024` and `NUM_ITER=100000` the execution takes 8 seconds. While [similar code written in HIP](https://github.com/illuhad/hipSYCL/files/8451133/hip_reproducer.txt) with the same parameters takes: 0.3s. 

Furthermore, it is very interesting that in case the USM interface is used the performance of hipSYCL is comparable to HIP (0.6s)
```
    //sycl::atomic_ref<float, sycl::memory_order::relaxed, sycl::memory_scope::device, sycl::access::address_space::global_space> ref(data[0]);

    q.submit([&](sycl::handler& h) {
        h.parallel_for<class axpy>(sycl::range<1>{NUM_THREADS}, [=] (sycl::id<1> it) {
                sycl::atomic_ref<float, sycl::memory_order::relaxed, sycl::memory_scope::device, sycl::access::address_space::global_space> ref(data[0]);
                float one = 1;
                for(int i = 0; i < NUM_ITER; ++i){
                        ref.fetch_add(one);
                }
        });
    });
```
It doesn't make a difference if the reference is created inside the kernel or outside. 

**Expected behavior**
I would expect efficient code to be generated for both the buffer-accessor and the USM case.

**Describe your setup**
* hipSYCL compiled from the current develop: 1046a78777a23ca75c6ea4e92291f1dbe36169ca
* hipSYCL installed on top of rocm 5.1.0
   * rocm-clang 14.0.0
   * rocm-clang 14.0.0
   * ROCm 5.1.0
* used invocation of syclcc: `syclcc -O3  -save-temps=obj  --hipsycl-targets=hip:gfx908 `

**Optional additional diagnostic information**

A look at the generated assembly explains the difference.
in the case of the buffer-accessor:
```
        v_mov_b32_e32 v3, s1
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_add_f32_e32 v0, 1.0, v1
        v_mov_b32_e32 v2, s0
        flat_atomic_cmpswap v0, v[2:3], v[0:1] glc
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_eq_u32_e32 vcc, v0, v1
        s_or_b64 s[2:3], vcc, s[2:3]
        v_mov_b32_e32 v1, v0
        s_andn2_b64 exec, exec, s[2:3]
        s_cbranch_execnz .LBB2_14
; %bb.15:                               ; %atomicrmw.end41
```
(This repeats multiple times)
while in the case of the USM:
```
        v_mov_b32_e32 v1, 1.0
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
        global_atomic_add_f32 v0, v1, s[0:1]
```
 

**Additional context**
Found when investigating performance issue of GROMACS with SYCL backend

**Comment by illuhad** [comment]
There's no difference in how `atomic_ref` is implemented and operates for those two cases. Maybe the amdgpu LLVM backend gets confused by additional C++ layers in the buffer-accessor model.
Can you try extracting a pointer from the accessor?
```c++
float* ptr = X.get_pointer();
sycl::atomic_ref<...> r{ptr[0]};

```

**Comment by illuhad** [comment]
This is what will be called in both cases:
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/include/hipSYCL/sycl/libkernel/generic/hiplike/atomic_builtins.hpp#L365

What clang turns that into is not up to us.

**Comment by sbalint98** [comment]
> Can you try extracting a pointer from the accessor?

Unfortunately, this does not resolve this issue for me

> What clang turns that into is not up to us.

Yes, I was afraid of that. I was hoping that you have a silver bullet solution to this problem. Do you think this might be related to how accessors are initialized in the kernel code? Do you think trying to add `__forcinline` to the `fetch_add` might help in this case?

**Comment by al42and** [comment]
Per my experience, Clang optimizes this into `global_atomic_add_f32` when the result is not used, because that's the limitation of floating-point atomics on GFX908. If it can not prove that the result is not used, it will put the cmpswap-loop.

For me, forcing inlining did not help much (or I was doing it wrong), so in GROMACS we decided to manually call `atomicAddNoRet`. Related: https://github.com/ROCm-Developer-Tools/hipamd/issues/19

**Comment by illuhad** [comment]
The functions here

https://github.com/illuhad/hipSYCL/blob/develop/include/hipSYCL/sycl/libkernel/atomic_builtins.hpp
https://github.com/illuhad/hipSYCL/blob/develop/include/hipSYCL/sycl/libkernel/generic/hiplike/atomic_builtins.hpp

are already marked as `HIPSYCL_BUILTIN` / `HIPSYCL_HIPLIKE_BUILTIN` which should include a `forceinline`. Only the `atomic_ref` class itself is not forceinline, but I would not expect this to make  a difference.

**Comment by sbalint98** [comment]
Thank you, @al42and, for the link. I couldn't find your workaround on the official GROMACS repo. Could you point me to the right branch? When `atomicAddNoRet` was used with hipSYCL source interop I could get more efficient instructions generated. However, I still couldn't get the amd-clang to generate efficient instructions with pure SYCL.   

I had a look at the LLVM IR, and I only found a few differences between the [USM](https://github.com/illuhad/hipSYCL/files/8472916/reproducer-usm-llvmir.txt) and [buffer accessor](https://github.com/illuhad/hipSYCL/files/8472917/reproducer-buffer-accessor-llvmir.txt) case:

buffer-accessor case:
```
define amdgpu_kernel void @_Z16__hipsycl_kernelIN7hipsycl4glue20complete_kernel_nameIZZ4mainENKUlRNS0_4sycl7handlerEE_clES5_E4axpyEEEvv(%class.anon addrspace(4)* nocapture readonly byref(%class.anon) align 8 %0, i64 %1, i64 %2, i1 %3) local_unnamed_addr #1 comdat {
  %.sroa.011.0..sroa_idx = getelementptr inbounds %class.anon, %class.anon addrspace(4)* %0, i64 0, i32 0, i32 0, i32 0, i32 0, i32 0, i64 0
  %.sroa.011.0.copyload = load i64, i64 addrspace(4)* %.sroa.011.0..sroa_idx, align 8
....
  %19 = inttoptr i64 %.sroa.011.0.copyload to float*
  %20 = atomicrmw fadd float* %19, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
  %21 = atomicrmw fadd float* %19, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
  %22 = atomicrmw fadd float* %19, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
...
```
USM case:
```
define amdgpu_kernel void @_Z16__hipsycl_kernelIN7hipsycl4glue20complete_kernel_nameIZZ4mainENKUlRNS0_4sycl7handlerEE_clES5_E4axpyEEEvv(float addrspace(1)* %0, i64 %1, i64 %2, i1 %3) local_unnamed_addr #1 comdat {
  %5 = addrspacecast float addrspace(1)* %0 to float*
...
  %19 = atomicrmw fadd float* %5, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
  %20 = atomicrmw fadd float* %5, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
  %21 = atomicrmw fadd float* %5, float 1.000000e+00 syncscope("agent-one-as") monotonic, align 4
```

Despite the similarity, the buffer-accessor IR is turned into CAS loops. I also tried adding the `-fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics` as suggested by the linked issue but that only helped with ROCm 5.0. To me the only major difference between the two IRs is the cast from int to pointer in the buffer-accessor case.

I investigated the performance with multiple rocm versions and different approaches for using atomic add in hipSYCL. On gfx908 using the flags only had an impact on the ROCm 5.0 case, otherwise, for the pure buffer-accessor examples, the same slow CAS loops were generated. Although the generated instructions were only aligned with pure HIP in the USM case, both using atomicAddNoRet and USM had approximately the same performance. 

<table class="tg">
<thead>
  <tr>
    <th class="tg-0pky" rowspan="2">gfx908(MI100) with hipSYCL latest </br> runtime in seconds </th>
    <th class="tg-0pky" colspan="2">4.5.2<br></th>
    <th class="tg-0pky" colspan="2">5.0</th>
    <th class="tg-0lax" colspan="2">5.1</th>
  </tr>
  <tr>
    <th class="tg-0lax">no<br>flags<br></th>
    <th class="tg-0lax">with<br>flags<br></th>
    <th class="tg-0lax">no<br>flags<br></th>
    <th class="tg-0lax">with<br>flags<br></th>
    <th class="tg-0lax">no<br>flags<br></th>
    <th class="tg-0lax">with<br>flags<br></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0pky">buffer-accessor</td>
    <td class="tg-0pky">8.4</td>
    <td class="tg-0lax">8.4</td>
    <td class="tg-0pky">8.3</td>
    <td class="tg-0lax">8.3</td>
    <td class="tg-0lax">8.3</td>
    <td class="tg-0lax">8.3</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-get-pointer</td>
    <td class="tg-0pky">8.49<br></td>
    <td class="tg-0lax">8.6</td>
    <td class="tg-0pky">8.5</td>
    <td class="tg-0lax">7.5</td>
    <td class="tg-0lax">8.2</td>
    <td class="tg-0lax">8.2</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicadd</td>
    <td class="tg-0pky">8.6</td>
    <td class="tg-0lax">8.1</td>
    <td class="tg-0pky">8.5</td>
    <td class="tg-0lax">8.2</td>
    <td class="tg-0lax">8.3</td>
    <td class="tg-0lax">8.3</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicaddnoret</td>
    <td class="tg-0pky">0.7</td>
    <td class="tg-0lax">0.7</td>
    <td class="tg-0pky">0.6</td>
    <td class="tg-0lax">0.6</td>
    <td class="tg-0lax">0.6</td>
    <td class="tg-0lax">0.6</td>
  </tr>
  <tr>
    <td class="tg-0pky">usm<br></td>
    <td class="tg-0pky">7.9</td>
    <td class="tg-0lax">7.8</td>
    <td class="tg-0pky">8.4</td>
    <td class="tg-0lax">0.6</td>
    <td class="tg-0lax">0.6</td>
    <td class="tg-0lax">0.6</td>
  </tr>
</tbody>
</table>

<table class="tg">
<thead>
  <tr>
    <th class="tg-0pky" rowspan="2">gfx908 with hipSYCL latest</br> instructions</th>
    <th class="tg-0pky" colspan="2">4.5.2<br></th>
    <th class="tg-0pky" colspan="2">5.0</th>
    <th class="tg-0pky" colspan="2">5.1</th>
  </tr>
  <tr>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0pky">buffer-accessor</td>
    <td class="tg-0pky" colspan="6" rowspan="3">flat_atomic_cmpswap<br>CAS<br>loop</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-get-pointer</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicadd</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicaddnoret</td>
    <td class="tg-0pky" colspan="6">mix of flat_atomic_cmpswap and<br>global_atomic_add_f32</td>
  </tr>
  <tr>
    <td class="tg-0pky">usm<br></td>
    <td class="tg-0pky" colspan="3">flat_atomic_cmpswap<br>CAS<br>loop</td>
    <td class="tg-0pky" colspan="3">global_atomic_add_f32</td>
  </tr>
</tbody>
</table>

Unfortunately, there was no MI200 card available to collect performance results. Nonetheless, I had a look at the assembly generated for the gfx90a target, and there are no significant improvements. The most notable difference is that the flags do not affect the generated instructions at all. 

<table class="tg">
<thead>
  <tr>
    <th class="tg-0pky" rowspan="2">gfx90a with hipSYCL latest </br> instructions </th>
    <th class="tg-0pky" colspan="2">4.5.2<br></th>
    <th class="tg-0pky" colspan="2">5.0</th>
    <th class="tg-0pky" colspan="2">5.1</th>
  </tr>
  <tr>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
    <th class="tg-0pky">no<br>flags<br></th>
    <th class="tg-0pky">with<br>flags<br></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0pky">buffer-accessor</td>
    <td class="tg-0pky" colspan="6" rowspan="3">flat_atomic_cmpswap<br>CAS<br>loop</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-get-pointer</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicadd</td>
  </tr>
  <tr>
    <td class="tg-0pky">buffer-accessor-source-interop-atomicaddnoret</td>
    <td class="tg-0pky" colspan="6">mix of flat_atomic_cmpswap and<br>global_atomic_add_f32</td>
  </tr>
  <tr>
    <td class="tg-0pky">usm<br></td>
    <td class="tg-0pky" colspan="4">flat_atomic_cmpswap<br>CAS<br>loop</td>
    <td class="tg-0pky" colspan="2">global_atomic_add_f32<br></td>
  </tr>
</tbody>
</table>
 
reproducers used:
[reproducer-usm.txt](https://github.com/illuhad/hipSYCL/files/8481212/reproducer-usm.txt)
[buffer-accessor.txt](https://github.com/illuhad/hipSYCL/files/8481213/buffer-accessor.txt)
[buffer-accessor-source-interop-atomicaddnoret.txt](https://github.com/illuhad/hipSYCL/files/8481214/buffer-accessor-source-interop-atomicaddnoret.txt)
[buffer-accessor-source-interop-atomicadd.txt](https://github.com/illuhad/hipSYCL/files/8481215/buffer-accessor-source-interop-atomicadd.txt)
[buffer-accessor-get-pointer.txt](https://github.com/illuhad/hipSYCL/files/8481216/buffer-accessor-get-pointer.txt)

**Comment by al42and** [comment]
> Thank you, @al42and, for the link. I couldn't find your workaround on the official GROMACS repo. Could you point me to the right branch? When atomicAddNoRet was used with hipSYCL source interop I could get more efficient instructions generated. 

The original hack was made last month: https://gitlab.com/gromacs/gromacs/-/compare/master...dardel-hackathon#4bce28c2b38888b0071aeba70c1b03a47db951d6_46_46

I only got around to making it a proper fix recently :) https://gitlab.com/gromacs/gromacs/-/merge_requests/2643

> To me the only major difference between the two IRs is the cast from int to pointer in the buffer-accessor case.

Could it be the missing `addrspacecast` in the buffer-accessor case? As I understand, optimization enabled with `munsafe-loop` checks the address spaces of the operands, so if it can not deduce it, it uses CAS loop.

**Comment by sbalint98** [comment]
As additional information: [the simple hip reproducer](https://github.com/illuhad/hipSYCL/files/8451133/hip_reproducer.txt) Generates fast assembly without using `atomicAddNoRet` and additional flags on rocm 5.1, 5.0, 4.5.2.

**Comment by illuhad** [comment]
hm.. currently we store a unique id of 64 bit integers inside accessors that is later cast to a pointer:
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/include/hipSYCL/glue/embedded_pointer.hpp#L111
I wonder if things would improve if we went the other way and stored a pointer that is reinterpreted as 64-bit ints on host...

**Comment by sbalint98** [comment]
Well, I had a look at this change, but unfortunately, it did not resolve the issue. The problem, in this case, seems to be that the memory is marked to be in addresspace(4) by LLVM which corresponds to constant memory. However, the [AMDGPU backend](https://github.com/llvm/llvm-project/blob/d46fa023caa2db5a9f1e21dd038bcb626261d958/llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp#L110-L118) only optimizes in case of local or global memory.

**Comment by illuhad** [comment]
@nilsfriess to investigate address space cast using clang attribute

**Comment by nilsfriess** [comment]
I tried to reproduce the examples and without additional flags, neither the USM nor the buffer accessor reproducer generate fast assembly (i.e., `global_atomic_add_f32` instructions).

When compiling with `-munsafe-fp-atomics` the USM code produces fast assembly, the buffer accessor does not (so I assume the results in the very first comment were compiled with this flag).

I now tried to change the following line to cast the pointer to a different address space:
https://github.com/illuhad/hipSYCL/blob/1046a78777a23ca75c6ea4e92291f1dbe36169ca/include/hipSYCL/glue/embedded_pointer.hpp#L127
I am actually not sure how this should be done, I have tried the following (maybe these are even the same things?):
- `return reinterpret_cast<T __attribute__((address_space(1)))*>(_uid.id[0]);`
- `return reinterpret_cast<__attribute__((address_space(1))) T*>(_uid.id[0]);`

However, when trying to recompile the buffer accessor reproducer, compilation fails with an error that says that casting to a different address space is not allowed. But as I said, I am not even sure if this is the correct way to cast.
